### PR TITLE
[Model][ROCm] Enable DeepSeek V4 Vision

### PR DESCRIPTION
**Pull request by AndreasKaratzas**
- Enable `DeepseekV4ForConditionalGeneration` on ROCm by moving the platform-neutral wrapper to `common/`, retaining the NVIDIA compatibility shim and unsupported XPU stub, and enabling ROCm registry, dummy-init, and tensor-schema paths.
- Build checkpoint mappings from the active text backend, preserving NVIDIA self-finalization while deferring and idempotently applying ROCm finalization after generic per-layer quantization.
- Route image sentinels through `bias_vl` in hash and regular ROCm MoE layers, expose local logits, and extend sparse-prefill visibility across complete image spans without changing text-only row width.
- This extends [PR #54566](https://github.com/vllm-project/vllm/pull/54566), merged as [`1356635d837c4ef002ec98c1a0296e7ff60be3c1`](https://github.com/vllm-project/vllm/commit/1356635d837c4ef002ec98c1a0296e7ff60be3c1), which introduced the checkpoint with a NVIDIA-only wrapper and non-CUDA guards.

AI assistance: OpenAI Codex assisted with merge resolution, source review, and regression testing.

### CODE REVIEW

File: vllm/models/deepseek_v4/amd/rocm.py [hunk 1]
```diff
@@ -172,12 +186,22 @@ def combine_topk_swa_indices(
...
 ) -> tuple[torch.Tensor, torch.Tensor]:
+    if (left_visible is None) != (right_visible is None):
+        raise ValueError("left_visible and right_visible must be provided together")
     topk_indices = topk_indices.reshape(topk_indices.shape[0], -1).contiguous()
     num_tokens = topk_indices.shape[0]
     num_reqs = seq_lens.shape[0]
+    has_image = left_visible is not None
+    # Keep the row shape fixed for a vision model even when a particular batch
+    # has no image. The gathered KV workspace needs no matching expansion: its
+    # query portion already contains each atomically-prefilled image span.
```
Comment (okorzh-amd) [inline vllm/models/deepseek_v4/amd/rocm.py]: This comment states the load-bearing precondition ("atomically-prefilled image span"), but ROCm never establishes it. is_mm_prefix_lm is True for the vision variant (transformers_utils/model_arch_config_convertor.py:606, whose own comment says it forces disable_chunked_mm_input) — yet that forcing exists only in vllm/platforms/cuda.py:320-331. RocmPlatform.check_and_update_config has no equivalent, so with chunked prefill a chunk boundary inside an image span sends these indices outside the gathered KV workspace. Please mirror the CUDA block in RocmPlatform, or hoist it into Platform.check_and_update_config so future backends inherit it.

File: vllm/models/deepseek_v4/amd/rocm.py [hunk 2]
```diff
@@ -131,7 +136,16 @@ def _combine_topk_swa_indices_kernel(
...
-        swa_len = tl.minimum(pos + 1, WINDOW_SIZE)
+        if HAS_IMAGE:
+            left = tl.load(left_visible_ptr + token_idx)
+            right = tl.load(right_visible_ptr + token_idx)
+        else:
+            left = 0
+            right = 0
+        left_add = tl.maximum(left - (WINDOW_SIZE - 1), 0)
+        swa_start = tl.maximum(pos - (WINDOW_SIZE - 1) - left_add, 0)
+        swa_end = pos + right + 1
```
Comment (okorzh-amd) [inline vllm/models/deepseek_v4/amd/rocm.py]: swa_end = pos + right + 1 reads forward of pos, and swa_start (line 146) collapses to span_start for a token deep inside a span — neither is clamped to gather_start. An APC prefix-cache hit landing mid-span under-runs the request's SWA region; --disable-chunked-mm-input does not fix this. Suggest swa_start = tl.maximum(swa_start, gather_start) and swa_end = tl.minimum(pos + right + 1, seq_len), recomputing swa_len from the clamped bounds. Note the same clamp looks absent from the shared common/ops/cache_utils.py kernel — if so this is pre-existing and affects CUDA too, and should be split into its own PR rather than blocking this one. Could you confirm?

File: tests/models/test_deepseek_v4_vl_rocm.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+from types import SimpleNamespace
+
+import pytest
+import torch
+from torch import nn
+
+from vllm.model_executor.models.utils import WeightsMapper
+
+pytestmark = pytest.mark.cpu_test
```
Comment (okorzh-amd) [inline tests/models/test_deepseek_v4_vl_rocm.py]: cpu_test is not accurate: three of the six tests import amd.model → amd/rocm.py:24 → vllm/platforms/rocm.py, whose module-level _GCN_ARCH = _get_gcn_arch() falls back to torch.cuda.get_device_properties("cuda").gcnArchName with no non-ROCm escape. The sibling tests/models/test_dspark_mla.py only gets away with a module-level model import because it imports an nvidia/ module. Separately, this file is named by no CI job — the buildkite lanes enumerate explicit filenames with no tests/models/*.py glob, and the -m cpu_test lanes are scoped to v1/core/multimodal/etc. Suggest a file-level skipif(not current_platform.is_rocm()) (matching tests/kernels/attention/test_rocm_triton_attn_dsv4.py:11) plus an entry in a ROCm lane, otherwise none of these tests can ever fail in CI.

File: tests/models/test_initialization.py [hunk 4]
```diff
@@ -160,7 +161,12 @@ def _initialize_kv_caches_v1(self, vllm_config):
         if model_arch == "WhisperForConditionalGeneration":
             m.setenv("VLLM_WORKER_MULTIPROC_METHOD", "spawn")
 
-        kwargs = {}
+        kwargs: dict[str, Any] = {}
+        if (
+            model_arch == "DeepseekV4ForConditionalGeneration"
+            and current_platform.is_rocm()
+        ):
+            kwargs["kv_cache_dtype"] = "fp8"
```
Comment (okorzh-amd) [inline tests/models/test_initialization.py]: This is the only per-arch KV-dtype override in the file, and it papers over a real usability gap rather than a test quirk: sparse_mla.py:47-51 advertises "auto" as a supported KV dtype, but attention.py:105 asserts kv_cache_dtype.startswith("fp8"). vllm serve deepseek-ai/DeepSeek-V4-Flash-Vision-Exp with default flags passes config validation and then dies on a bare AssertionError with no remediation text. Please either coerce auto → fp8_ds_mla on that path, or raise a ValueError naming --kv-cache-dtype fp8 and drop "auto" from supported_kv_cache_dtypes — then this override can go away.

File: vllm/models/deepseek_v4/amd/model.py [hunk 5]
```diff
@@ -294,6 +310,8 @@ def forward(
     ) -> torch.Tensor:
         if self.gate.tid2eid is not None and input_ids is None:
             raise ValueError("DeepSeek V4 hash MoE routing requires input_ids.")
+        if self.gate.bias_vl is not None and input_ids is None:
```
Comment (okorzh-amd) [inline vllm/models/deepseek_v4/amd/model.py:313]: bias_vl is allocated in every MoE layer when vision_n_layers > 0, including the MTP block, and amd/mtp.py:159 hardcodes input_ids=None. Enabling spec-decode on a vision checkpoint therefore hits this ValueError deep inside MoE forward. Refusing is correct (the mapper deliberately drops mtp. weights), but this should be rejected at config time with a message saying MTP/DSpark is unsupported for the vision variant.

File: tests/models/multimodal/processing/test_tensor_schema.py [hunk 6]
```diff
@@ -165,9 +165,9 @@ def test_model_tensor_schema(model_id: str):
         )
 
     if model_id == "deepseek-ai/DeepSeek-V4-Flash-Vision-Exp" and not (
-        current_platform.is_cuda()
+        current_platform.is_cuda() or current_platform.is_rocm()
```
Comment (Isotr0py) [inline tests/models/multimodal/processing/test_tensor_schema.py]:
```suggestion
        current_platform.is_cuda_alike()
```

File: tests/models/test_registry.py [hunk 7]
```diff
@@ -65,9 +65,9 @@ def test_registry_imports(model_arch):
         pytest.skip("HY V4 is only supported on CUDA")
 
     if model_arch == "DeepseekV4ForConditionalGeneration" and not (
-        current_platform.is_cuda()
+        current_platform.is_cuda() or current_platform.is_rocm()
```
Comment (Isotr0py) [inline tests/models/test_registry.py]: ditto

**Review by Isotr0py — COMMENTED** [review]
Look reasonable. But would like to let @tjtanaa to check the rocm implementation.

File: tests/models/test_registry.py (hunk 7 above)
Comment (AndreasKaratzas) [reply tests/models/test_registry.py]: Done :)

File: vllm/models/deepseek_v4/amd/rocm.py (hunk 1 above)
Comment (AndreasKaratzas) [reply vllm/models/deepseek_v4/amd/rocm.py]: Fixed in the latest push. I mirrored CUDA's multimodal prefix-LM guard in RocmPlatform.check_and_update_config, so ROCm now forces disable_chunked_mm_input.

File: vllm/models/deepseek_v4/amd/rocm.py (hunk 2 above)
Comment (AndreasKaratzas) [reply vllm/models/deepseek_v4/amd/rocm.py]: Right, the shared common/ops/cache_utils.py path has the same pre-existing missing bounds. For this PR, I applied the gather_start/seq_len clamps to the new ROCm kernel and added an APC-mid-span regression test. I left the shared CUDA/common change out of this PR, cause I think it can be addressed separately.

File: tests/models/test_deepseek_v4_vl_rocm.py (hunk 3 above)
Comment (AndreasKaratzas) [reply tests/models/test_deepseek_v4_vl_rocm.py]: Removed the cpu_test marker, added a file-level ROCm guard, and included the test file in the Basic Models lane mirrored onto MI300.

File: tests/models/test_initialization.py (hunk 4 above)
Comment (AndreasKaratzas) [reply tests/models/test_initialization.py]: Right, I removed the test-only override and made the packed layout resolve auto to fp8_ds_mla, while explicit unsupported dtypes now produce an actionable ValueError.

File: vllm/models/deepseek_v4/amd/model.py (hunk 5 above)
Comment (AndreasKaratzas) [reply vllm/models/deepseek_v4/amd/model.py:313]: `input_ids=None` indeed causes the vision-routing failure, but I don’t think configuration-time rejection is necessary. The target mapper drops mtp.* because the separate DeepSeekV4MTP draft model remaps and loads those weights itself. NVIDIA MTP already forwards input_ids, and AMD DSpark does as well. I updated AMD MTP to do the same and added a regression test.

**Comment by AndreasKaratzas** [comment]
DeepSeek-V4-Flash-Vision-Exp on 4× AMD gfx950 per run (ROCm, TP4/EP, FP8 KV, AITER, 1,024 batched tokens, graph mode NONE; #55107 `141ab249ecf1` + #55095 `c6b3c36d87b4`). Full test sets with greedy decoding; GSM8K uses vLLM's 5-shot CI protocol (256-token output limit).

| Benchmark | Baseline | DSpark (3 tokens) | Request errors (B / D) | Truncations (B / D) |
| --- | --- | --- | --- | --- |
| OCRBench | **825 / 1,000 (82.5%)** | **822 / 1,000 (82.2%)** | 0 / 0 | 0 / 0 |
| GSM8K | **1,243 / 1,319 (94.2%)** | **1,239 / 1,319 (93.9%)** | 0 / 0 | 3 / 3 |

OCRBench category accuracy uses the official scoring rule, thinking disabled, and a 4,096-token output limit.

| Category | Baseline | DSpark (3 tokens) |
| --- | --- | --- |
| Regular Text Recognition | 47 / 50 (94.0%) | 47 / 50 (94.0%) |
| Irregular Text Recognition | 46 / 50 (92.0%) | 45 / 50 (90.0%) |
| Artistic Text Recognition | 47 / 50 (94.0%) | 47 / 50 (94.0%) |
| Digit String Recognition | 33 / 50 (66.0%) | 34 / 50 (68.0%) |
| Non-Semantic Text Recognition | 45 / 50 (90.0%) | 45 / 50 (90.0%) |
| Handwriting Recognition | 32 / 50 (64.0%) | 32 / 50 (64.0%) |
| **Text Recognition (subtotal)** | **250 / 300 (83.3%)** | **250 / 300 (83.3%)** |
| Scene Text-centric VQA | 175 / 200 (87.5%) | 176 / 200 (88.0%) |
| Doc-oriented VQA | 177 / 200 (88.5%) | 173 / 200 (86.5%) |
| Key Information Extraction | 175 / 200 (87.5%) | 175 / 200 (87.5%) |
| Handwritten Mathematical Expression Recognition | 48 / 100 (48.0%) | 48 / 100 (48.0%) |

DSpark acceptance is aggregated over each full run using Prometheus counter deltas with fixed verification (adaptive verification disabled).

| Metric | OCRBench | GSM8K |
| --- | --- | --- |
| Mean acceptance length (tokens/forward) | **3.00** | **3.25** |
| Draft acceptance (accepted / drafted) | **66.5%** (33,820 / 50,844) | **75.1%** (81,209 / 108,159) |
| Per-position acceptance (1 / 2 / 3) | 84.8% / 67.2% / 47.6% | 90.4% / 76.3% / 58.5% |

**Review by tjtanaa — APPROVED** [review]
LGTM. Let check the Cuda ci.
